### ISSUE TITLE
http-Proxy settings are ignored

### ISSUE THREAD

**Issue by BerengarWLehr** [maths/moodle-qtype_stack#1021]
Moodle's proxy-settings e.g. `proxyhost`, `proxyport` etc. are not taken into account when trying to connect via curl.

**Comment by aharjula** [comment]
A bit more context, please.

What exactly are you aiming to do with curl? This sounds a lot like an API branch issue, or are we talking about some part of STACK in particular? Iframes, JSXGraphs or some such? Maybe server maxima connections, although those might not be nice to push through the general proxy anyway.

**Comment by BerengarWLehr** [comment]
The problem occurs in https://github.com/maths/moodle-qtype_stack/blob/ec5cc8e6dffb2b915c2fa682870e2c1ef8af396a/stack/cas/connector.server.class.php#L39-L54 when a raw curl request is created without considering proxy settings. One can easily fix this locally by adding the following line, but this is far away from a good solution.

```php
curl_setopt($request, CURLOPT_PROXY, $CFG->proxytype . '://' . $CFG->proxyhost . ':' . $CFG->proxyport);
```
Probably moodle's http_client class built upon guzzle (/lib/classes/http_client.php) should be used but I'm not really qualified for such suggestion.

**Comment by aharjula** [comment]
Ok. So you would like the Maxima server connection to go through the same proxy as all other Moodles external traffic? I might have a problem with that, as most server Maxima setups have their servers in internal networks and don't tend to need proxies for those or at least not the same proxies as for general external communication (typically these have their own load-balancers as their proxies). Applying that proxy setting might break a large set of installs where such internal servers are used, as it might be that that traffic could not bounce back from the proxy due to the internal servers being firewalled to only allow the Moodle nodes to communicate with them.

I do know there are cloud versions of goemaxima and someone has surely made MaximaPool cloud setups as well, and those could need the proxy, but those are not really the designed use pattern. I do wonder if we should have another connector class so that people could pick that if they wish/need to use that proxy.

@sangwinc could we have a `connector.server-proxy.class.php` style of a thing that would be the same as `connector.server.class.php` but with that extra line added?

**Comment by BerengarWLehr** [comment]
Knowing that there always will be edge cases that are not covered, I think in most cases the `proxybypass` configuration should help to not proxy internal servers.

**Comment by aharjula** [comment]
Indeed, with the correct settings, everything works, especially if we were to use the http_client instead of curl directly. However, simply switching proxying on for those that have not had to define any `proxybypass` settings would cause some issues, issues that I would rather avoid by simply creating the connector as is and creating another connector for those that need proxying where all these settings could be applied, after that they can simply change the used connector (the platform setting) setting and all will work.

Such a simple change in a long-established behaviour, no matter how well documented will still trigger many a support ticket, especially in a system as complex as server maximas are to set up. And I for one, really do not want to start debugging other people's proxies, if at all avoidable.

**Comment by aharjula** [comment]
Ok. The details i.e. why I think we need to have a separate old version and a Moodle proxy setting following version.

So, some Moodle installs use HTTP proxy to route the external requests from within the system. This is one way to secure things as if you apply firewalling and stop all traffic leaving the system (internals of the system not from the user's browser) that does not leave through the proxy, you can then further filter it in the proxy if need be. However, all proxying tends to cost, and typically, you do not apply it to internal stuff. In particular, this may be of use when setting up some sort of an exam system where traffic is specially filtered.

In the past, the assumption has been that the MaximaPool/server maxima is always an internal thing and does not have any special proxies or at least not proxies controlled by the Moodle proxy settings. Now, this issue requests that the Moodle settings would apply also to that connection, and it does make some sense in some cases. If we were to follow the Moodle proxy settings, then one could even use them to configure it so that proxying would not apply to these servers (`proxybypass`). What worries me is that simply starting to follow these settings could cause a situation as follows:
 1.  One could break one's server config by suddenly starting to pass traffic to a proxy that could not then pass it back to the server maxima. This could happen during the update of STACK if the proxy settings of that Moodle are already defined or it could happen at a later point when someone decides to turn proxying on in that Moodle. That latter case is problematic as the breakage might not be immediately visible due to CAS-caches.
 2. Even if we document that there is a change, how many will really notice that, and will understand to set the `proxybypass` setting if it applies in their setup? Server Maximas are already hard enough, and as this applies to those who have actually fought and built their own internal setups instead of using some cloud solution this really worries me.

Well, we can do this with minimal duplicate code if we manage to build the proxy setting flowing version and implement the original as version of that where `proxybypass` always matches.


As a side comment about the security of those cloud Maxima server implementations, I really do not believe that are safe. After all, you can input absolutely anything into them, and I really doubt that they cannot be broken out of and easily DOSsed. Internal Maxima servers at least have the benefit of limiting where the command comes from and if those come from real STACK install, then STACK limits what goes into them.

@sangwinc The proxy settings have no security implications that matter, and you do not need to understand them. The only issue is the change of behaviour that depends on the configuration of the Moodle in which STACK is installed in, leading to debugging issues if that configuration causes trouble where it previously had no role.

In general, my view is that following those proxy settings is an edge case, I do consider the set of systems where that is relevant to be small and most notably smaller than the set of systems that will actively need to avoid using them. It will also be a case where ILIAS will need to revert some code unless they have some proxy settings of their own they wish to map to this.

In the end, we have two options:
 1. We add a new "platform"/connection for those that wish to use their Moodle proxy settings.
 2. We simply follow Moodle proxy settings in the current connector, and those that it causes trouble for will need to figure out what happened and how to fix it.

**Comment by nwp90** [comment]
Hi all - coming here from a separate discussion about stack_include being unusable for us due to its not using configured Moodle proxies.

From a security standpoint, it is absolutely basic that Moodle provides a proxy configuration, and *everything* within Moodle should be using it. It is equally true that where Moodle provides a standard means of performing an action, especially where there are potential security implications, *everything* that needs to perform that action should use the standard Moodle-provided means to do so (barring a /really good/ reason not to).

Given that, I think the only question is "how do we get there from here".

As @aharjula says, we don't want to break existing installs without providing plenty of (preferably unavoidable) warnings first.

The right way to do this kind of thing is generally to add an option to do things the right way, which by default does not modify the behaviour of existing installs but would be active for new installs. Initially this is mentioned in documentation. In a subsequent version, you would add a deprecation warning for those still using the old setup (perhaps with a link to documentation on using the new setup), and eventually the option to do it the old way is removed.

I have to disagree with @aharjula about the security implications of proxy settings; proxies are generally used/required for security-related reasons, and a recent Moodle stable update contained a security fix for the code that handles the proxy settings - and broke many previously-valid configurations in the process. "You do not need to understand them" is correct iff you are using the Moodle-provided means of fetching remote content.

The complication with stack_include is that it uses PHP's file_get_contents, which it seems can get either local or remote content. It's conceivable that people could be using it to include files located on the Moodle server, but outside of Moodle's knowledge/control. If it does currently work to include local content, any fix would have to make sure it was still possible to fetch local content, and this would probably require checking the provided string to see whether it looked like a URL (ideally in an identical manner to whatever file_get_contents does internally, but that's unlikely to be possible).

For the connector class, I suspect the best way to fix it would be to modify it to use Moodle's client library, but to override the proxy setting if the "old way" is desired - no need to duplicate the whole thing.

Given that we need this fixed for stack_include ASAP, I guess we're happy to do the work if we can agree on the basic requirements.

**Comment by aharjula** [comment]
Ok, the `file_get_contents` is a major issue and will be fixed for the next release, thank you for publicly bringing it up here, let's hope no one uses that to its full power, @sangwinc  this might require some mailing around, for the major users... Expect a commit within a couple of hours. Luckily it is abusable only from the author's side.

As to the proxy settings being a security issue please describe any attack that would cause security issues using the fact that we have no proxy settings in use, to me there simply is no security issue there that requires ASAP handling. In particular, nothing at the scale of the hole you just gave up with the `file_get_contents`. To me, it sounds like security and routing are being confused here and you are using the word security as a way to push a particular routing method forcibly forward. If people ask for routing please ask for it and don't turn my mind to the security paranoia mode where I object to everything unless it is actually proven to improve security in all cases.

**Comment by aharjula** [comment]
Now if you really want to use Moodles proxying and for inclusions you can do so by replacing this bit of code:

https://github.com/maths/moodle-qtype_stack/blob/c2a49eeaae9827920b8cc23a2e91e20f28414153/vle_specific.php#L282-L289

Or more relevantly the `file_get_contents` call there that only receives HTTP and HTTPS URLs at that point of logic. Note that this would be a Moodle-specific portion of logic and is therefore now externalised to the `vle_specific.php`-scripts file.

**Comment by sangwinc** [comment]
Thanks Matti,
I was looking through this as well in parallel and I appreciate you pushing your commit.
Chris

**Comment by nwp90** [comment]
Doh. Was so focused on what we're trying to achieve that file_get_contents being a security issue locally didn't even begin to cross my mind :-/

Re security issues - failure to observe proxy settings will allow access to hosts that an admin would expect to be blocked by any ACLs on the proxy. Admins rightly expect Moodle proxy settings to be observed by everything in Moodle.

If this kind of thing didn't matter, then https://moodle.org/mod/forum/discuss.php?d=449640#p1807042 wouldn't be getting a CVE.

Failure to use Moodle's supplied curl helper in a plugin - https://www.cvedetails.com/cve/CVE-2022-45152/ 

FWIW, the ASAP comes from our need - in our setup if we don't use the proxy, we don't get access.

**Comment by EJMFarrow** [comment]
Hi all,

Thanks for all your investigations of this. I've created a fix for the original issue with a new platform/ connector class as suggested which handles the various Moodle proxy settings.

https://github.com/maths/moodle-qtype_stack/compare/dev...iss1021

It also throws up a warning in the healthcheck if the 'server' platform is being used, Moodle proxyhost is set and the maxima server is not in Moodle's proxybypass list.

I've tested locally with Docker, etc, but is anyone able to test the change with a more realistic setup?
